core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,904 @@
+      x0: 0.6350000000000001,
+      y0: 0.3933333333333333,
+      xLabel: 0.7250000000000001,
+      yLabel: 0.43333333333333335,
+      id: 'State7',
+      label: 'State7',
+      height: 0.08,
+      width: 0.18000000000000002,
+      reachableFromEnd: false
+    },
+    State8: {
+      depth: 3,
+      offset: 0,
+      reachable: true,
+      x0: 0.07250000000000001,
+      y0: 0.5266666666666667,
+      xLabel: 0.1625,
+      yLabel: 0.5666666666666667,
+      id: 'State8',
+      label: 'State8',
+      height: 0.08,
+      width: 0.18000000000000002,
+      reachableFromEnd: true
+    },
+    State9: {
+      depth: 1,
+      offset: 0.5,
+      reachable: true,
+      x0: 0.185,
+      y0: 0.26,
+      xLabel: 0.275,
+      yLabel: 0.30000000000000004,
+      id: 'State9',
+      label: 'State9',
+      height: 0.08,
+      width: 0.18000000000000002,
+      reachableFromEnd: true
+    },
+    Orphaned: {
+      depth: 5,
+      offset: 0,
+      reachable: false,
+      x0: 0.07250000000000001,
+      y0: 0.7933333333333333,
+      xLabel: 0.1625,
+      yLabel: 0.8333333333333333,
+      id: 'Orphaned',
+      label: 'Orphaned',
+      height: 0.08,
+      width: 0.18000000000000002,
+      reachableFromEnd: false
+    }
+  };
+  let links1: GraphLink[] = [
+    {
+      source: 'State1',
+      target: 'State1'
+    },
+    {
+      source: 'State1',
+      target: 'State2'
+    },
+    {
+      source: 'State1',
+      target: 'State3'
+    },
+    {
+      source: 'State1',
+      target: 'State5'
+    },
+    {
+      source: 'State1',
+      target: 'State6'
+    },
+    {
+      source: 'State1',
+      target: 'State7'
+    },
+    {
+      source: 'State1',
+      target: 'State8'
+    },
+    {
+      source: 'State1',
+      target: 'State9'
+    },
+    {
+      source: 'State2',
+      target: 'State4'
+    },
+    {
+      source: 'State3',
+      target: 'State4'
+    },
+    {
+      source: 'State9',
+      target: 'State8'
+    },
+    {
+      source: 'State8',
+      target: 'State4'
+    },
+  ];
+
+  let links2: GraphLink[] = [
+    {
+      source: 'State1',
+      target: 'State1'
+    },
+    {
+      source: 'State1',
+      target: 'State2'
+    },
+    {
+      source: 'State1',
+      target: 'State3'
+    },
+    {
+      source: 'State2',
+      target: 'State4'
+    },
+    {
+      source: 'State3',
+      target: 'State4'
+    },
+  ];
+
+  beforeEach(() => {
+    sgls = TestBed.inject(StateGraphLayoutService);
+  });
+
+  it('should create adjacency lists', () => {
+    let nodes: GraphNodes = {
+      State1: 'State1',
+      State2: 'State2',
+      State3: 'State3',
+      State4: 'State4',
+    };
+
+    let adjacencyLists = {
+      State1: ['State2', 'State3'],
+      State2: ['State4'],
+      State3: ['State4'],
+      State4: []
+    };
+
+    expect(sgls.getGraphAsAdjacencyLists(nodes, links2))
+      .toEqual(adjacencyLists);
+  });
+
+  it('should get indentation levels', () => {
```

Comment:
Think about the behaviour and requirements for this. You don't just want the service to return indentation levels, you want it to return the correct ones. So, write out the exact requirements for this service, and then have tests that cover each requirement.

Reply:
Added another test